### PR TITLE
Addressed spectral warn not fail on CI and updated oasdiff

### PR DESCRIPTION
**Pull request by ruvyas**
## Changes
- Fixed bug where `appsettings.json` was not being applied when executing MSBuild task
- Updated `Oasdiff` to `1.9.6` -> This fixes relative file path issue
- Fixed `Spectral` cmd not failing on `warn` errors

### CODE REVIEW

**Comment by PrincessMadMath** [comment]
If there is a diff in the OasDiff should it output a warning so it could make the CI pipelines fail? Locally it's ok they diverge until the sourced-control spec file is updated but once it's ready to merge with main they should have no diff no?

File: src/Workleap.OpenApi.MSBuild/msbuild/tools/Workleap.OpenApi.MSBuild.targets [hunk 1]
```diff
@@ -15,6 +15,9 @@
   <!-- We can only analyze the OpenAPI spec after the project has been built -->
   <Target Name="ValidateOpenApi" DependsOnTargets="ResolveProjectReferences" AfterTargets="Build">
     <PropertyGroup>
+      <!-- The path of the ASP.NET Core project being built -->
+      <StartupAssemblyPath Condition="'$(StartupAssemblyPath)' == ''">$(MSBuildProjectDirectory)\$(OutputPath)</StartupAssemblyPath>
```
Comment (PrincessMadMath) [inline src/Workleap.OpenApi.MSBuild/msbuild/tools/Workleap.OpenApi.MSBuild.targets:19]: We could update the internal documentation reference: https://gsoftdev.atlassian.net/wiki/spaces/TEC/pages/3839557876/MSBuild+Parameters

File: src/Workleap.OpenApi.MSBuild/OasdiffManager.cs [hunk 2]
```diff
@@ -62,7 +59,7 @@ private async Task DecompressDownloadedFileAsync(string oasdiffFileName, Cancell
             return;
         }
         
-        var exitCode = await this._processWrapper.RunProcessAsync("tar", new[] { "-xzf", $"{pathToCompressedFile}", "-C", $"{this._oasdiffDirectory}" }, cancellationToken);
+        var exitCode = await this._processWrapper.RunProcessAsync("tar", new[] { "-xzf", $"{pathToCompressedFile}", "-C", $"{this._oasdiffDirectory}" }, cancellationToken: cancellationToken);
```
Comment (PrincessMadMath) [inline src/Workleap.OpenApi.MSBuild/OasdiffManager.cs:62]: Why the change of specifying the parameter name?

File: src/WebApiDebugger/Controllers/WeatherForecastController.cs [hunk 3]
```diff
@@ -4,6 +4,7 @@ namespace WebApiDebugger.Controllers;
 
 [ApiController]
 [Route("[controller]")]
+[Produces("application/json")]
```
Comment (PrincessMadMath) [inline src/WebApiDebugger/Controllers/WeatherForecastController.cs:7]: Was this for testing purpose?

File: src/Workleap.OpenApi.MSBuild/SpectralManager.cs [hunk 4]
```diff
@@ -86,7 +84,7 @@ private async Task GenerateSpectralReport(string spectralExecutePath, string swa
             await this.AssignExecutePermission(spectralExecutePath, cancellationToken);
         }
 
-        var exitCode = await this._processWrapper.RunProcessAsync(spectralExecutePath, new[] { "lint", swaggerDocumentPath, "--ruleset", rulesetUrl, "--format", "html", "--output.html", htmlReportPath }, cancellationToken);
+        var exitCode = await this._processWrapper.RunProcessAsync(spectralExecutePath, new[] { "lint", swaggerDocumentPath, "--ruleset", rulesetUrl, "--format", "html", "--format", "pretty", "--output.html", htmlReportPath, "--fail-severity=warn" }, cancellationToken: cancellationToken);
```
Comment (PrincessMadMath) [inline src/Workleap.OpenApi.MSBuild/SpectralManager.cs:87]: (same nit-pick comment about CancellationToken): why specifying the parameter name? Seems to add verbosity

**Comment by PrincessMadMath** [comment]
Would it be possible to add delimitator logs for which tools is running

Currently it look like this:
image


Could be nice to have logs delimated with something like:

****************** OasDiff ******************
{...}
****************** Spectral ******************
{...}

File: src/Workleap.OpenApi.MSBuild/SpectralManager.cs (hunk 4 above)
Comment (PrincessMadMath) [reply src/Workleap.OpenApi.MSBuild/SpectralManager.cs:87]: (also applicable in the couple others places)

File: src/Workleap.OpenApi.MSBuild/ValidateOpenApiTask.cs [hunk 5]
```diff
@@ -26,13 +30,14 @@ public sealed class ValidateOpenApiTask : CancelableAsyncTask
 
     protected override async Task<bool> ExecuteAsync(CancellationToken cancellationToken)
     {
+        var reportsPath = Path.Combine(this.OpenApiToolsDirectoryPath, "reports");
         var loggerWrapper = new LoggerWrapper(this.Log);
-        var processWrapper = new ProcessWrapper(this.OpenApiToolsDirectoryPath);
+        var processWrapper = new ProcessWrapper(this.StartupAssemblyPath);
```
Comment (PrincessMadMath) [inline src/Workleap.OpenApi.MSBuild/ValidateOpenApiTask.cs:35]: Do we want the startup path or the output? Let say they have a appsettings.json but they doesn"t copy on build, the tools will work but it won't be the same behavior as a real release.

File: src/Workleap.OpenApi.MSBuild/ValidateOpenApiTask.cs [hunk 6]
```diff
@@ -75,6 +87,27 @@ protected override async Task<bool> ExecuteAsync(CancellationToken cancellationT
...
+    private bool CheckIfBaseSpecExists()
+    {
+        foreach (var file in this.OpenApiSpecificationFiles)
+        {
+            if (File.Exists(file))
+            {
+                continue;
+            }
+
+            this.Log.LogWarning(
```
Comment (PrincessMadMath) [inline src/Workleap.OpenApi.MSBuild/ValidateOpenApiTask.cs:99]: Should we copy it for them? Initialize their current spec file! We could write a comment "we have initialize your spec file, it will now be used as the base comparaison"

Comment (PrincessMadMath) [reply src/Workleap.OpenApi.MSBuild/ValidateOpenApiTask.cs:99]: I think it could be a better UX

**Review by PrincessMadMath — APPROVED** [review]

**Comment by ruvyas** [comment]
@PrincessMadMath We can gather that data from user testing. I guess the counter argument is, OasDiff's job is just to highlight changes. If we fail it might block CI unless teams specifically wants to do this

File: src/Workleap.OpenApi.MSBuild/OasdiffManager.cs (hunk 2 above)
Comment (ruvyas) [reply src/Workleap.OpenApi.MSBuild/OasdiffManager.cs:62]: I like to name `CancellationToken` param. Makes it easier to extend in future if we add a default parameter.

File: src/WebApiDebugger/Controllers/WeatherForecastController.cs (hunk 3 above)
Comment (ruvyas) [reply src/WebApiDebugger/Controllers/WeatherForecastController.cs:7]: Yup. Just making sure debug project doesn't generate any diff from Oasdiff

File: src/Workleap.OpenApi.MSBuild/ValidateOpenApiTask.cs (hunk 5 above)
Comment (ruvyas) [reply src/Workleap.OpenApi.MSBuild/ValidateOpenApiTask.cs:35]: As far as I know .net looks for `appsettings.json` in build output folder by default. You can of course change this behavior but seems like an edge case. I would like to see an example of it.

File: src/Workleap.OpenApi.MSBuild/ValidateOpenApiTask.cs (hunk 6 above)
Comment (ruvyas) [reply src/Workleap.OpenApi.MSBuild/ValidateOpenApiTask.cs:99]: I don't think we should. It is a one-time task so not every dev will have to do go through this. Plus I think developers would want freedom of where their specs are saved. In which case they will have to "copy + paste" anyway. Maybe we can add this for future enhancement or prioritize based on user-feedback

**Comment by ruvyas** [comment]
> Would it be possible to add delimitator logs for which tools is running
> 
> Currently it look like this: image
> 
> Could be nice to have logs delimated with something like:
> 
> ****************** OasDiff ****************** {...} ****************** Spectral ****************** {...}

They both currently run simultaneously. It would look weird and incoherent.

**Comment by ruvyas** [comment]
> If there is a diff in the OasDiff should it output a warning so it could make the CI pipelines fail? Locally it's ok they diverge until the sourced-control spec file is updated but once it's ready to merge with main they should have no diff no?

We can gather that data from user testing. I guess the counter argument is, OasDiff's job is just to highlight changes. If we fail it might block CI unless teams specifically wants to do this

**Comment by PrincessMadMath** [comment]
Can we print the current file we are validating when there is multiplpe doc generation: 
image

File: src/Workleap.OpenApi.MSBuild/SpectralManager.cs [hunk 7]
```diff
@@ -86,7 +84,7 @@ private async Task GenerateSpectralReport(string spectralExecutePath, string swa
             await this.AssignExecutePermission(spectralExecutePath, cancellationToken);
         }
 
-        var exitCode = await this._processWrapper.RunProcessAsync(spectralExecutePath, new[] { "lint", swaggerDocumentPath, "--ruleset", rulesetUrl, "--format", "html", "--output.html", htmlReportPath }, cancellationToken);
+        var exitCode = await this._processWrapper.RunProcessAsync(spectralExecutePath, new[] { "lint", swaggerDocumentPath, "--ruleset", rulesetUrl, "--format", "html", "--format", "pretty", "--output.html", htmlReportPath, "--fail-severity=warn" }, cancellationToken);
```
Comment (Gcaya) [inline src/Workleap.OpenApi.MSBuild/SpectralManager.cs:87]: It works if you have twice the --format argument?

File: src/Workleap.OpenApi.MSBuild/ValidateOpenApiTask.cs [hunk 8]
```diff
@@ -61,6 +67,12 @@ protected override async Task<bool> ExecuteAsync(CancellationToken cancellationT
             await installOasdiffTask;
 
             var generateOpenApiDocsPath = (await swaggerManager.RunSwaggerAsync(this.OpenApiSwaggerDocumentNames, cancellationToken)).ToList();
+            
+            if (!this.CheckIfBaseSpecExists())
```
Comment (Gcaya) [inline src/Workleap.OpenApi.MSBuild/ValidateOpenApiTask.cs:71]: Can this check be done even earlier so we early exit before doing any work or are there certain actions we need to take first before we can do the check?

File: src/Workleap.OpenApi.MSBuild/SpectralManager.cs (hunk 7 above)
Comment (ruvyas) [reply src/Workleap.OpenApi.MSBuild/SpectralManager.cs:87]: Yup. Oasdiff is weird. It doesn't work if you specify multiple formatter with just one switch

File: src/Workleap.OpenApi.MSBuild/ValidateOpenApiTask.cs (hunk 8 above)
Comment (ruvyas) [reply src/Workleap.OpenApi.MSBuild/ValidateOpenApiTask.cs:71]: Well, I think we should still generate spec(s) for devs from code even if they don't have base spec. They can use our generated spec as a baseline for thei "base" spec.

File: src/Workleap.OpenApi.MSBuild/OasdiffManager.cs [hunk 9]
```diff
@@ -42,14 +40,13 @@ public async Task RunOasdiffAsync(IEnumerable<string> openApiSpecFiles, IEnumera
...
         {
-            this._loggerWrapper.LogMessage($"Starting Oasdiff comparison with {specFile}.");
+            this._loggerWrapper.LogMessage($"Starting Oasdiff comparison with {baseSpecFile}.");
             
-            var fileName = Path.GetFileName(specFile);
-            var baseSpecRelativePath = Path.GetRelativePath(this._openApiToolsDirectory, specFile);
-            var newSpecRelativePath = Path.GetRelativePath(this._openApiToolsDirectory, generatedOpenApiSpecFilesList.First(x => x.Contains(fileName)));
-            await this._processWrapper.RunProcessAsync(oasdiffExecutePath, new[] { "diff", baseSpecRelativePath, newSpecRelativePath, "--exclude-elements", "description,examples,title,summary", "-f", "text" }, cancellationToken);
+            var fileName = Path.GetFileName(baseSpecFile);
+            var newSpecRelativePath = generatedOpenApiSpecFilesList.First(x => x.Contains(fileName));
```
Comment (Gcaya) [inline src/Workleap.OpenApi.MSBuild/OasdiffManager.cs:48]: Is it possible that this one could be null? If it's possible, should we error out?

**Review by Gcaya — APPROVED** [review]

**Review by PrincessMadMath — APPROVED** [review]
